File: librarymanager/librarymanager-be/src/main/java/ru/spring5/elena/librarymanager/domain/entity/ReaderEntity.java
package ru.spring5.elena.librarymanager.domain.entity;

import lombok.*;
import org.hibernate.annotations.Formula;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

import static javax.persistence.GenerationType.IDENTITY;
/*
 * Класс для  хранения инфы о читателях библиотеки
 */

@Entity
@Table(name = "READERS")
@Getter
@NoArgsConstructor
//@RequiredArgsConstructor
@Setter
public class ReaderEntity implements Serializable {
    /**
     *
     */
    private static final long serialVersionUID = 993948553969252992L;

    @Id
    @GeneratedValue(strategy = IDENTITY) //уникальный ID, генерируется сам в БД
    @Column(name = "READER_ID")
    @NotNull
    @Setter(AccessLevel.NONE)
    private Long readerId;

    @Version
    @Column(name = "VERSION")
    private int version; //версия, инкрементится при редактировании записи

    @NotEmpty(message = "{validation.firstname.notempty.message}")
    @Size(min = 3, max = 60, message = "{validation.firstname.size.message}")
    @Column(name = "FIRST_NAME")
    private String firstName; //имя

    @NotEmpty(message = "{validation.surname.notempty.message}")
    @Size(min = 1, max = 60, message = "{validation.surname.size.message}")
    @Column(name = "SURNAME")
    private String surname; //фамилия

    @Size(min = 0, max = 60, message = "{validation.patronymic.Size.message}")
    @Column(name = "PATRONYMIC")
    private String patronymic; //отчество

    @NotNull(message = "{validation.birthday.NotEmpty.message}")
    @Temporal(TemporalType.DATE)
    @Column(name = "BIRTHDATE")
    private Date birthdate; //дата рождения, чтобы камасутру мелким не выдать

    @Temporal(TemporalType.DATE)
    @Column(name = "REG_DATE")
    private Date regDate;

    @Size(min = 0, max = 12, message = "{validation.telephone.Size.message}")
    @Column(name = "TELEPHONE")
    private String telephone; //телефон, чтобы присылать смс о готовых заказах и задолжностях

    @NotEmpty(message = "{validation.email.NotEmpty.message}")
    @Size(min = 4, max = 200, message = "{validation.email.Size.message}")
    @Column(name = "EMAIL")
    private String email; //электронная почта. Уникальное поле, используется для авторизации

    @OneToMany(mappedBy = "reader", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter(AccessLevel.NONE)
    private Set<OrderEntity> orders = new HashSet<OrderEntity>();

    @OneToMany(mappedBy = "reader", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter(AccessLevel.NONE)
    private Set<BookUsingEntity> bookUsings = new HashSet<BookUsingEntity>();

    @Formula("(SURNAME || ' ' || FIRST_NAME || ' ' || PATRONYMIC)")
    @Transient
    private String fio;

}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/ReaderEntity.kt
package ru.elenacherev.librarymanager.domain.entity

import org.hibernate.annotations.ColumnDefault
import org.hibernate.annotations.Formula
import java.util.*
import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.OneToMany
import javax.persistence.Table
import javax.persistence.Temporal
import javax.persistence.TemporalType
import javax.persistence.Transient
import javax.persistence.Version
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

/*
 * Класс для  хранения инфы о читателях библиотеки
 */
@Entity
@Table(name = "READERS")
data class ReaderEntity(

    @Column(name = "READER_ID", nullable = false, updatable = false)
    @GeneratedValue
    @Id
    @get:NotNull
    var readerId: UUID? = null,

    //версия, инкрементится при редактировании записи
    @Version
    @Column(name = "VERSION", nullable = false)
    var version: Int = 0,

    //имя
    @Column(name = "FIRST_NAME")
    @NotEmpty
    @Size(min = 3, max = 60)
    var firstName: String,

    //фамилия
    @Column(name = "SURNAME")
    @NotEmpty
    @Size(min = 1, max = 60)
    var surname: String? = null,

    //отчество
    @Column(name = "PATRONYMIC")
    @Size(min = 0, max = 60)
    var patronymic: String? = null,

    //дата рождения, чтобы камасутру мелким не выдать
    @Column(name = "BIRTHDATE")
    @NotNull
    @Temporal(TemporalType.DATE)
    var birthdate: Date,

    @Column(name = "REG_DATE")
    @Temporal(TemporalType.DATE)
    var regDate: Date,

    //телефон, чтобы присылать смс о готовых заказах и задолжностях
    @Column(name = "TELEPHONE")
    @Size(min = 0, max = 12)
    var telephone: String? = null,

    //электронная почта. Уникальное поле, используется для авторизации
    @Column(name = "EMAIL")
    @NotEmpty
    @Size(min = 4, max = 200)
    var email: String,

    @OneToMany(mappedBy = "reader", cascade = [CascadeType.ALL], orphanRemoval = true)
    val orders: MutableSet<OrderEntity> = mutableSetOf(),

    @OneToMany(mappedBy = "reader", cascade = [CascadeType.ALL], orphanRemoval = true)
    val bookUsings: MutableSet<BookUsingEntity> = mutableSetOf()
) {
    @Formula("(SURNAME || ' ' || FIRST_NAME || ' ' || PATRONYMIC)")
    @Transient
    lateinit var fio: String
}
